File: ova-backend/src/main/java/com/example/ovabackend/service/RatingService.java
package com.example.ovabackend.service;

import com.example.ovabackend.model.Ova;
import com.example.ovabackend.model.Rating;
import com.example.ovabackend.model.User;
import com.example.ovabackend.repository.OvaRepository;
import com.example.ovabackend.repository.RatingRepository;
import com.example.ovabackend.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class RatingService {

    @Autowired
    private RatingRepository ratingRepository;

    @Autowired
    private OvaRepository ovaRepository;

    @Autowired
    private UserRepository userRepository;

    public List<Rating> getAllRatings() {
        return ratingRepository.findAll();
    }

    public Optional<Rating> getRatingById(Long id) {
        return ratingRepository.findById(id);
    }

    public List<Rating> getRatingsByOvaId(Long ovaId) {
        return ratingRepository.findByOvaIdOrderByFechaCreacionDesc(ovaId);
    }

    public Double getAverageRatingByOvaId(Long ovaId) {
        Double average = ratingRepository.getAverageRatingByOvaId(ovaId);
        return average != null ? average : 0.0;
    }

    public Rating createRating(Long ovaId, Long userId, Rating rating) {
        Ova ova = ovaRepository.findById(ovaId)
                .orElseThrow(() -> new RuntimeException("OVA no encontrado con id: " + ovaId));
        
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("Usuario no encontrado con id: " + userId));

        // Check if user already rated this OVA
        Optional<Rating> existingRating = ratingRepository.findByOvaIdAndUserId(ovaId, userId);
        if (existingRating.isPresent()) {
            throw new RuntimeException("El usuario ya ha calificado este OVA");
        }

        rating.setOva(ova);
        rating.setUser(user);
        return ratingRepository.save(rating);
    }

    public Rating updateRating(Long id, Rating ratingDetails) {
        Rating rating = ratingRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Calificación no encontrada con id: " + id));

        rating.setCalificacion(ratingDetails.getCalificacion());
        rating.setComentario(ratingDetails.getComentario());

        return ratingRepository.save(rating);
    }

    public void deleteRating(Long id) {
        ratingRepository.deleteById(id);
    }
}

File: ova-frontend/src/app/services/rating.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Rating {
  id?: number;
  calificacion: number;
  comentario?: string;
  fechaCreacion?: string;
}

export interface RatingAverage {
  average: number;
}

@Injectable({
  providedIn: 'root',
})
export class RatingService {
  private apiUrl = 'http://localhost:8080/api/ratings';

  constructor(private http: HttpClient) {}

  getAllRatings(): Observable<Rating[]> {
    return this.http.get<Rating[]>(this.apiUrl);
  }

  getRatingById(id: number): Observable<Rating> {
    return this.http.get<Rating>(`${this.apiUrl}/${id}`);
  }

  getRatingsByOvaId(ovaId: number): Observable<Rating[]> {
    return this.http.get<Rating[]>(`${this.apiUrl}/ova/${ovaId}`);
  }

  getAverageRatingByOvaId(ovaId: number): Observable<RatingAverage> {
    return this.http.get<RatingAverage>(`${this.apiUrl}/ova/${ovaId}/average`);
  }

  createRating(ovaId: number, userId: number, rating: Rating): Observable<Rating> {
    return this.http.post<Rating>(`${this.apiUrl}/ova/${ovaId}/user/${userId}`, rating);
  }

  updateRating(id: number, rating: Rating): Observable<Rating> {
    return this.http.put<Rating>(`${this.apiUrl}/${id}`, rating);
  }

  deleteRating(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
